File: leetcode/sort_list.js
/* Sort a linked list in O(n log n) time using constant space complexity.
 * Time Complexity: O(n lg n)
 * Space Complexity: O(1)
 */

/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var sortList = function(head) {
    var dummy = new ListNode(null);
    dummy.next = head;
    for (var len = 0, slow = dummy; slow.next; ++len)
        slow = slow.next;
    for (var merge_size = 1; merge_size < len; merge_size *= 2) {
        for (var run_start = dummy; run_start.next; ) {
            var pf = run_start.next, ps = run_start.next;
            slow = run_start.next;
            for (var fast = run_start.next, i = 0; i < merge_size; ++i) {
                pf = slow;
                slow = slow.next;
                if (!slow)
                    break;
                if (fast) {
                    ps = fast;
                    fast = fast.next;
                }
                if (fast) {
                    ps = fast;
                    fast = fast.next;
                }
            }
            if (!slow)
                break;
            pf.next = null;
            ps.next = null;
            var first_list = run_start.next, second_list = slow;
            for (; first_list || second_list; run_start = run_start.next) {
                if ((first_list && !second_list)
                 || (first_list && first_list.val < second_list.val)) {
                     var temp = first_list.next;
                     run_start.next = first_list;
                     first_list = temp;
                 } else {
                     temp = second_list.next;
                     run_start.next = second_list;
                     second_list = temp;
                 }
            }
            run_start.next = fast;
        }
    }
    return dummy.next;
};

File: lintcode/sort_list.py
"""
Definition of ListNode
class ListNode(object):

    def __init__(self, val, next=None):
        self.val = val
        self.next = next
"""
class Solution:
    """
    @param head: The first node of the linked list.
    @return: You should return the head of the sorted linked list,
                  using constant space complexity.
    """
    def sortList(self, head):
        dummy = ListNode(None, head)
        len, slow = 0, dummy
        while slow.next:
            len += 1
            slow = slow.next
        merge_size = 1
        while merge_size < len:
            run_start = dummy # node before two lists
            while run_start.next:
                second_list = first_list = pf = ps = slow = fast =run_start.next
                for _ in xrange(merge_size):
                    pf, slow = slow, slow.next
                    if not slow: break
                    if fast: ps, fast = fast, fast.next
                    if fast: ps, fast = fast, fast.next
                if not slow: break
                pf.next, ps.next, second_list = None, None, slow
                while first_list or second_list: # merge two lists
                    if (first_list and not second_list) or (first_list
                                  and first_list.val < second_list.val):
                        run_start.next, first_list = first_list, first_list.next
                    else:
                        run_start.next,second_list=second_list,second_list.next
                    run_start = run_start.next
                run_start.next = fast
            merge_size *= 2
        return dummy.next
